### ISSUE TITLE
[COT] `query()` method: allow extensibility by 3rd parties

### ISSUE THREAD

**Issue by jorgeatorres** [woothemes/woocommerce#33879]
3rd party code should be able to extend query vars (and the query itself) to allow querying of other tables/fields.

**Comment by jorgeatorres** [comment]
- Review https://github.com/woocommerce/woocommerce/wiki/wc_get_orders-and-WC_Order_Query.
- `woocommerce_order_data_store_cpt_get_orders_query` filter.

**Comment by webdados** [comment]
I need the `woocommerce_order_data_store_cpt_get_orders_query` filter, or a replacement for it, on the new HPOS tables for several of my payment gateway plugins.

Usage example, of one of the pluginxs:
- During checkout an API call is made to the payment gateway, which returns a unique payment request ID that is stored to the order meta
- The customer gets a push notification on his phone and accepts the payment
- The payment gateway makes a webhook/API call to the website stating that the payment request with that unique ID was actually paid for
- My plugin queries the database for pending payment orders with that unique ID and sets it as paid

I do know meta is not the best place to search, but the flexibility needs to be there, or else databases will become a nightmare of different lookup tables implemented by each and every plugin that needs to search by its own custom meta .

**Comment by jorgeatorres** [comment]
Hi @webdados!

Thanks for letting us know of your use case. The implementation of `wc_get_orders()` (in HPOS) is much more flexible than the current one, including compatibility with the `meta_query` syntax that [`WP_Query` uses](https://developer.wordpress.org/reference/classes/wp_meta_query/), which makes me think you could make use of this instead of having to override things with a filter similar to `woocommerce_order_data_store_cpt_get_orders_query`.

For example, you could do something like this:

```php
$orders_associated_to_id = wc_get_orders(
	array(
		'meta_query' => array(
			array(
				'key'   => 'the_order_meta_key_you_are_looking_for',
				'value' => 'the_unique_id'
			)
		),
		'limit' => 1,
	)
);
```

I believe this would be enough flexibility for what you need, but please let me know if that's not the case. We definitely want to understand the needs of our devs for this iteration of the querying API.

By the way, the above `meta_query` support (as well as support for date queries) is already part of HPOS. You can even take a look at some examples [in our tests](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php).
That said, I'm going to prioritize adding some good examples to our [HPOS upgrade guide](https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book).

**Comment by webdados** [comment]
I wasn't aware of that compatibility. 
I guess I'll just have to check if HPOS is being used or not and run a different query: the old one using filters, or the new one using the `meta_query` directly on `wc_get_orders()`.

I was thinking of a scenario in which filters already in place should work seamlessly.

**Comment by jorgeatorres** [comment]
> I wasn't aware of that compatibility.

Yes. As I said, we need to add some examples to the upgrade guide and document the new features better.

> I was thinking of a scenario in which filters already in place should work seamlessly.

We do want to "backport" certain query features to the posts datastore, so maybe it'll work seamlessly even without the filters, but I'm not sure if that will make it into the next release.

I'll make sure to comment here once things have been merged to keep you in the loop!

**Comment by webdados** [comment]
> The implementation of `wc_get_orders()` (in HPOS) is much more flexible than the current one, including compatibility with the `meta_query` syntax that [`WP_Query` uses](https://developer.wordpress.org/reference/classes/wp_meta_query/), which makes me think you could make use of this instead of having to override things with a filter similar to `woocommerce_order_data_store_cpt_get_orders_query`.

I that's the case, the exact same filter could be used to transform old-style query args to HPOS query args.
On `OrdersTableDataStore.php`, in the `query` method, adding this before actually performing the query, does the job:
`//Make CPT woocommerce_order_data_store_cpt_get_orders_query filter work for order tables
$query_vars = apply_filters( 'woocommerce_order_data_store_cpt_get_orders_query', $query_vars, $query_vars, $this );`

Just to be clear, before this:
`try {
	$query = new OrdersTableQuery( $query_vars );
}`

This is just a simple test to check if the implementation works, which it does. It should be done in a more elegant way and throwing a deprecation notice, of course.

I'm saying this because in a plugin in which I added 13 entries, with the `woocommerce_order_data_store_cpt_get_orders_query` filter for custom fields to be searchable, I found myself creating a translation helper function that is basically the same as the function I did for the mentioned filter.

**Comment by clifgriffin** [comment]
Would be lovely to update the documentation on this.

Juggling HPOS/non-HPOS considerations with order queries with meta_queries is quite painful.

**Comment by davidmosterd** [comment]
I think the new flexibility is great, but I feel I miss the option to change the SQL directly. Not everything can be caught in an array I think..., or I have not figured out the code.

If I, for example, want to join the order items table to see if the amount of items is > 3, or has VAT, or whatever kind of scenario that 's' does not accommodate. How would we do that with HPOS now? I see no way of hooking into the query other than using arrays on predefined tables/ mappings.

Also not sure this is the right ticket, it's basically the `build_query` method on the `OrdersTableQuery`, but I figured it the same-ish issue.

**Comment by jorgeatorres** [comment]
Hi everyone!

We recently merged #36625 which introduced hooks `woocommerce_orders_table_query_clauses` and `woocommerce_orders_table_query_sql` allowing modification of the different clauses that make up the query and the final SQL respectively.

I believe that should cover the scenarios discussed in this issue, but feel free to re-open if I'm missing something. We're looking at ways to extend the querying mechanism with new operators/query types, but in the mean time, this should allow anyone to write code similar to what was possible in previous implementation.

Thanks everyone for the input!
